0011-CIR-trust-distribution.md

```diff
@@ -0,0 +1,116 @@
+
+- Implement an active trust distribution mechanism for Dapr in Kubernetes that responds to updates.
+- Enable root rotation with no downtime.
+
+### Non-Goals
+
+- Sentry implements CA root rotation.
+- Implement external CA support.
+- Implement Dapr trust federation.
+
+### Current Shortfalls
+
+Trust distribution is the act of propagating trust data to enable secure communication or networking between peers.
+In the case of Dapr, this involves propagating PEM encoded CA bundle files to clients and servers in the cluster, which are then used to authenticate peers over TLS.
+Today, their are two methods of CA deployment to Dapr; either Sentry generated or provided by the Dapr cluster administrator.
+From the prospective of trust distribution, these two modes are functionally the same as they both result in the `dapr-trust-bundle` ConfigMap containing the CA bundle, and the `dapr-trust-bundle` Secret containing the issuer certificate chain.
+From here, trust distribution for the control plane occurs by the Operator, Placement, and Injector services reading from the mounted `dapr-trust-bundle` ConfigMap.
+Trust distribution for Daprds occurs from the Injector patching the Daprd container with an environment variable containing the CA Bundle originating from the `dapr-trust-bundle` ConfigMap.
+
+The problem with the current strategy is that once trust is distributed once (the `dapr-trust-bundle` ConfigMap and Secret is populated), the root of trust cannot change in the cluster.
+This is because trust bundles are only injected to Dapr containers at Pod creation time.
+Trust anchors are also set as environment variables whose values are static for the entire duration of a unix process, meaning they cannot be dynamically updated, for example in the event of CA root rotation.
+Today, Dapr pods will have to be restarted in order to pick-up a new trust bundle.
+It is also undefined and untested as to whether the control plane components will successfully pick-up a new trust bundle during execution; though whether they can is irrelevant as Daprds do not also support this feature.
+
+## Solution
+
+### Trust Distributor
+
+It is paramount that the entity that conducts the trust distribution is separate from the entity that issues identities from that root of trust, in Dapr's case this is Sentry.
+This is because trust distribution must happen out of band of identity issuance.
+An analogous to this is roots of trust of the Internet are delivered via the computers Operating System or Internet Browser, rather than fetching them from DNS servers themselves.
+Similarly for example, asset SHA hashes should be downloaded from a separate source then the assert server themselves.
+Decoupling these roles also has the benefit of improving separation of concerns between responsibilities from the identity issuer, and the trust distributor.
+
+Trust distribution will be conducted by the Operator and written to the ConfigMap `dapr-root-ca.crt` in all Namespaces.
```

Comment:
Can this be written only to the dapr-system namespace? Eventually we will want to support multiple Dapr control planes in a cluster, and this seems to go the other direction

Reply:
I think we can tackle this with a `namespaced` mode of the trust distributor. Not writing to every namespace prevents trust being distributed at all.

Reply:
What if a new namespace is added? Will we be watching for new namespaces?

Reply:
Yeah exactly- I have implemented similar/same mechanisms in the past:
https://github.com/cert-manager/trust-manager/blob/main/pkg/bundle/sync.go
https://github.com/cert-manager/istio-csr/blob/main/pkg/controller/configmap.go